lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -13,548 +13,1055 @@
+    auto instances = symbolTable.getSymbolUses(op, circuitOp);
+    if (!instances || std::distance(instances.getValue().begin(),
+                                    instances.getValue().end()) == 0) {
+      op->emitOpError() << "is marked as a GrandCentral '" << msg
+                        << "', "
+                           "but is never instantiated";
+      return None;
     }
+    if (std::distance(instances.getValue().begin(),
+                      instances.getValue().end()) != 1) {
+      auto diag = op->emitOpError() << "is marked as a GrandCentral '" << msg
+                                    << "', but "
+                                       "it is instantiated more than once";
+      for (auto instance : instances.getValue())
+        diag.attachNote(instance.getUser()->getLoc())
+            << "parent is instantiated here";
+    }
+    return cast<InstanceOp>((*(instances.getValue().begin())).getUser());
+  };
 
-    // If the interface already exists, don't create it.
-    if (interfaces.count(bundle.defName))
-      return true;
-
-    // Create the interface.  This will be populated later.
-    interfaces[bundle.defName] =
-        builder.create<sv::InterfaceOp>(circuitOp->getLoc(), bundle.defName);
-
-    return true;
+  /// Walk the circuit and extract all information related to scattered
+  /// Grand Central annotations.  This is used to populate: (1) the
+  /// companionIDMap, (2) the parentIDMap, and (3) the leafMap.
+  /// Annotations are removed as they are discovered and if they are not
+  /// malformed.
+  removalError = false;
+  circuitOp.walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op)
+        .Case<RegOp, RegResetOp, WireOp, NodeOp>([&](auto op) {
+          AnnotationSet::removeAnnotations(op, [&](Annotation annotation) {
+            if (!annotation.isClass(
+                    "sifive.enterprise.grandcentral.AugmentedGroundType"))
+              return false;
+            auto maybeID = getID(op, annotation);
+            if (!maybeID)
+              return false;
+            leafMap.insert({maybeID.getValue(), op.getResult()});
+            return true;
+          });
+        })
+        // TODO: Figure out what to do with this.
+        .Case<InstanceOp>([&](auto op) {
+          AnnotationSet::removePortAnnotations(op, [&](unsigned i,
+                                                       Annotation annotation) {
+            if (!annotation.isClass(
+                    "sifive.enterprise.grandcentral.AugmentedGroundType"))
+              return false;
+            op.emitOpError()
+                << "is marked as an interface element, but this should be "
+                   "impossible due to how the Chisel Grand Central API works";
+            removalError = true;
+            return false;
+          });
+        })
+        .Case<MemOp>([&](auto op) {
+          AnnotationSet::removeAnnotations(op, [&](Annotation annotation) {
```

Comment:
nested lambdas.  oh boy.

Reply:
It's lambdas [all the way down](https://en.wikipedia.org/wiki/Lambda_calculus).

Reply:
Do you have any thoughts on how to better factor this?  These could be changed to private methods on `GrandCentralPass`, but I'm not sure if that helps readability.